tools/autograd/derivatives.yaml

```diff
@@ -2173,6 +2173,16 @@
 - name: linear(Tensor input, Tensor weight, Tensor? bias=None) -> Tensor
   input, weight, bias: linear_backward(input, grad, weight, grad_input_mask)
 
+- name: linear_backward(Tensor self, Tensor grad_output, Tensor weight, bool[3] output_mask) -> (Tensor, Tensor, Tensor)
+  self: "grads[1].defined() ? grad_output.mm(grads[1]) : Tensor()"
+  grad_output: "grads[0].defined() ?
+                (
+                 grads[1].defined() ? grads[0].mm(weight.mT()) + self.mm(grads[1].mT())
```

Comment:
I am concerned about the performance of these layers. @qqaatw , can you please benchmark these layers and check the performance ?

Reply:
This is higher order gradient support where I think most model/training won't access it. If the performance impact is significant, we can maybe branch on `GradMode::is_enabled()` like the below example for fused and infinitely differentiable paths, but I need some time to dig into those so would be inclined to get this PR merged first to unblock some users.

```
- name: logit(Tensor self, float? eps=None) -> Tensor
  self: "GradMode::is_enabled() ? infinitely_differentiable_logit_backward(grad, self, eps) : logit_backward(grad, self, eps)"
  result: auto_element_wise
```